content/blog/performant-elm-keyed/index.md

```diff
@@ -0,0 +1,155 @@
+---
+title: Performant Elm, Part II, Html Keyed
+date: "2019-10-07T12:00:00.000Z"
+description: Understand how Html Keyed works and how to use it.
+---
+
+This is the second post in a series which explores optimization techniques
+to build performant Elm applications. But first, let's take a bit of a
+detour and talk about [Saint Thomas](https://en.wikipedia.org/wiki/Thomas_the_Apostle).
+
+Saint Thomas was one of twelve apostles of Jesus and he is also known as
+_Doubting Thomas_. Why is that? Because of one specific episode in the New
+Testament. In the Gospel of John it is said that when he heard that Jesus
+had risen from the dead and appeared to the other apostles, he said:
+
+> Except I shall see on his hands the print of the nails, and put my finger
+> into the print of the nails, and thrust my hand into his side, I will not
+> believe.
+
+But when Jesus appeared later and invited Thomas to touch his wounds and
+behold him, Thomas believed and exclaimed "My Lord and my God". At which point Jesus said:
+
+> Thomas, because thou hast seen me, thou hast believed: blessed are
+> they that have not seen, and yet have believed.
+
+Well, my point about optimization techniques in general is: be more like
+Saint Thomas, aka Doubting Thomas. Don't believe what the documentation
+says, but instead go and see for yourself. Then, _believe_.
+
+So first of all I would recommend reading [the first post in this
+series](../performant-elm), which shows you how to use your browser
+profiling tools to perform repeatable measurements.
+
+## What is Html Keyed
+
+The wonderful [Elm
+Guide](https://guide.elm-lang.org/optimization/keyed.html) tells us that
+`Html.Keyed` is particularly helpful for lists of data in your interface
+that must support **insertion**, **removal**, and **reordering**.
+
+Imagine you have a list of words in ascending order: now you change to sort
+order to be descending. The virtual DOM algorithm will then:
+
+- Compare the previous first word with the current first word
+- Compare the previous second word with the current second word
+- ...
+
+As you can tell, the algorithm will have to compare each and every one of
+our words! Instead, it could have done a much faster job and rearranged the
+nodes in the DOM to appease our new sort order.
+
+So the basic idea of `Html.Keyed` is that you associate a `String` to your
+`Html msg` so that the virtual DOM diffing algorithm knows in advance which
+nodes have to be compared.
+
+## A working example
+
+Let's unleash our inner Doubting Thomas and test out these assumptions. In
+most UNIX systems there is a dictionary of words in
+`/usr/share/dict/words`, so I've copied more or less 3000 of those and
+wrote a <a href="https://ellie-app.com/6RtqCZC7Bzva1" target="_blank">small
+Ellie app</a> to display them. It looks something like this:
+
+![Ellie example](./ellie-example.png)
+
+To simulate a slighly more complex HTML structure, I've broken each word
```

Comment:
I might suggest using a smaller dataset. Right now going from not keyed -> keyed is going from very very slow to just very slow. You could demonstrate your point better if you made it fast.